crates/ruff_db/src/file_system/vendored.rs

```diff
@@ -0,0 +1,353 @@
+    fn read_zipfile(mut zipfile: ZipFile) -> Result<String> {
+        let mut buffer = String::with_capacity(zipfile.size() as usize);
+        zipfile.read_to_string(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+impl FileSystem for VendoredFileSystem {
+    fn exists(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        self.vendored_archive().lookup_path(&normalized).is_ok()
+            || (!normalized.has_trailing_slash()
+                && self
+                    .vendored_archive()
+                    .lookup_path(&normalized.with_trailing_slash())
+                    .is_ok())
+    }
+
+    fn is_directory(&self, path: &FileSystemPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        if let Ok(zipfile) = self.vendored_archive().lookup_path(&normalized) {
+            return zipfile.is_dir();
+        }
+        if normalized.has_trailing_slash() {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalized.with_trailing_slash())
+            .is_ok_and(|zipfile| zipfile.is_dir())
+    }
+
+    fn is_file(&self, path: &FileSystemPath) -> bool {
+        if path.as_str().ends_with('/') {
+            return false;
+        }
+        self.vendored_archive()
+            .lookup_path(&normalize_vendored_path(path))
+            .is_ok_and(|zipfile| zipfile.is_file())
+    }
+
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+
+        let is_dir = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => zipfile.is_dir(),
+            Err(err) => {
+                if normalized.has_trailing_slash() {
+                    return Err(err);
+                }
+                let mut archive = self.vendored_archive();
+                let lookup_result = archive.lookup_path(&normalized.with_trailing_slash());
+                if let Ok(zipfile) = lookup_result {
+                    zipfile.is_dir()
+                } else {
+                    return Err(err);
+                }
+            }
+        };
+
+        let file_type = if is_dir {
+            FileType::Directory
+        } else {
+            FileType::File
+        };
+
+        Ok(Metadata {
+            revision: FileRevision::zero(),
+            permissions: Some(Self::READ_ONLY_PERMISSIONS),
+            file_type,
+        })
+    }
+
+    fn read(&self, path: &FileSystemPath) -> Result<String> {
+        let normalized = normalize_vendored_path(path);
+        let mut archive = self.vendored_archive();
+        let lookup_error = match archive.lookup_path(&normalized) {
+            Ok(zipfile) => return Self::read_zipfile(zipfile),
+            Err(err) => err,
+        };
+        if normalized.has_trailing_slash() {
```

Comment:
What's the retry logic here? I get the impression that we do some trial and error here to read the file and that feels off. Can't we just return an error and ask callers to construct the right path instead of guessing paths?

Reply:
This is point (2) that I described in the PR description. For other file systems we use in Ruff, `file_system.exists("stdlib")` will return `true` if there exists a directory called `stdlib`. But without this retry logic, that will return `false` for the vendored filesystem, because zip archives use trailing slashes internally to distinguish between files and directories in the archive. So if there's a `stdlib` directory, then (without this retry logic) `file_system.exists("stdlib/")` would return `true`, but `file_system.exists("stdlib")` would return `false`.

I wanted to try to abstract over that difference in my implementation here, as I think it would be a bug magnet for paths to work differently for the vendored file system compared to how they work with other file systems we're using elsewhere in Ruff.

Reply:
...but this doesn't make any sense, because you can't read a directory, you can only read a file 🤦‍♂️